proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,221 @@
+
+NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
+This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
+`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
+Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.
+This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.
+
+#### Restore summary
+
+When restore is run at normal or detailed verbosity, NuGet outputs detected nuget.config files and package sources (feeds) used.
+Audit sources should be added.
+
+```diff
+ NuGet Config files used:
+     C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config
+
+ Feeds used:
+     https://internal.corp.contoso.com/packages/nuget/index.json
++
++Audit sources used:
++    https://api.nuget.org/v3/index.json
+```
+
+The "Audit sources used" section should ideally list sources where a VDB was successfully obtained.
+This depends on the feasibility of implementation.
+For example, if two package sources are defined, and no audit sources are defined by the *nuget.config* file, but only one of the two sources has a VDB, then the restore summary will only list the one source with a VDB under "Audit sources used".
+
+If `NuGetAudit` is disabled for all projects in the solution, the audit sources used section of the summary should be omitted.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+Not using `packageSources` for VDB when there's at least one source in `auditSources` is a performance optimization.
+Since we expect customers to use `auditSources` when their `packageSources` don't already provide a VDB, having restore spend time making HTTP requests to the package sources is unlikely to provide any benefit.
+While downloading the VDB from audit sources is likely to take more time than verifying that package sources don't have a vulnerability resource, if any package source is slow, downloading the VDB from the audit source might complete first.
+Additionally, in the case where NuGet doesn't otherwise need to communicate with any package source, checking the package sources for a VDB might trigger a credential provider.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+Customers who appear functionally network isolated from nuget.org (for example, by blocking nuget.org at the firewall) will still not be able to use this feature.
```

Comment:
After reading this and [this](https://learn.microsoft.com/dotnet/core/tools/dotnet-restore#audit-for-security-vulnerabilities), it's my (incorrect?) understanding that NuGetAudit currently requires NuGet.org as a package source, that even if I use my own package source that provides the V3 `VulnerabilityResource` NuGetAudit won't work.  Is that true?

I suspect no, it isn't true, that any package source that implements the V3 protocol sufficiently (including the `VulnerabilityResource` resource) could serve as a VDB source (as a package source) in .NET 8 SDK today.  This would be consistent with the V3 design:  any V3 package source can pick and choose which resources they want to support.  The docs (linked above) don't do a good job explaining that NuGet.org is recommended because it's the only package source we know that implements the `VulnerabilityResource` resource, and that any package source that implements the resource could be used in addition to or instead of NuGet.org.

```suggestion
Customers who appear functionally network isolated from nuget.org (for example, by blocking nuget.org at the firewall) will only be able to use this feature if their `auditSources` are accessible.
```

Reply:
> Is that true?

No, as you suspect, Audit works with any source that supports the vulnerability info resource. I guess the person who added that info to the dotnet docs misunderstood or was provided incorrect information. We'll get the docs updated.